sklearn/cluster/_kmeans.py

```diff
@@ -2138,9 +2138,18 @@ def fit(self, X, y=None, sample_weight=None):
             # Perform the iterative optimization until convergence
             for i in range(n_steps):
                 # Sample a minibatch from the full dataset
-                minibatch_indices = random_state.randint(0, n_samples, self._batch_size)
+                minibatch_indices = random_state.choice(
+                    n_samples,
+                    self._batch_size,
+                    p=sample_weight / np.sum(sample_weight),
+                    replace=True,
+                )
 
                 # Perform the actual update step on the minibatch data
+                # Note, I am not sure how sample weights are used here
+                # So left it in, it seems like the weight sums are updated using
+                # sample weights so need some help here to understand the
+                # _minibatch_update_dense/sparse code
```

Comment:
Like in KBinsDiscretizer, we should not use sample weights after using them for weighted sampling.
So let's pass an array of ones as sample weights here.

> I am not sure how sample weights are used here
   So left it in, it seems like the weight sums are updated using
   sample weights so need some help here to understand

weight_sum is the sum of weights of all points belonging to each clusters. It's used to track clusters where there are very few points (more precisely points that add up to a small weight) and reassign them to a different cluster.

I don't think we have to modify `_mini_batch_step`. It's still useful that it handles sample weights because it's also used in partial_fit where there is no sampling so sample weights must be passed.

Reply:
> I don't think we have to modify _mini_batch_step. It's still useful that it handles sample weights because it's also used in partial_fit where there is no sampling so sample weights must be passed.

That's a very good point indeed.